internal/services/notification/webhooksreceiver.go

```diff
@@ -0,0 +1,187 @@
+// Copyright 2023 Sorint.lab
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package notification
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"sync"
+	"testing"
+
+	"github.com/gorilla/mux"
+	"github.com/rs/zerolog"
+
+	"agola.io/agola/internal/testutil"
+	"agola.io/agola/internal/util"
+)
+
+type webhooksReceiver struct {
+	ctx     context.Context
+	cancel  context.CancelFunc
+	started bool
+
+	log zerolog.Logger
+
+	t *testing.T
+
+	mu sync.Mutex
+
+	listenAddress string
+	exposedURL    string
+
+	webhooks *webhooks
+}
+
+type webhooks struct {
+	webhooks []*webhook
+	mu       sync.Mutex
+}
+
+type webhook struct {
+	payload   []byte
+	signature string
+}
+
+func (ws *webhooks) getWebhooks() ([]*webhook, error) {
+	ws.mu.Lock()
+	defer ws.mu.Unlock()
+
+	retVal := make([]*webhook, len(ws.webhooks))
+	for i, w := range ws.webhooks {
+		retVal[i] = &webhook{payload: w.payload, signature: w.signature}
```

Comment:
If your intention was to create a deep copy, this isn't right. See how it was done from the integration tests webhook receivers.